### ISSUE TITLE
Detection of type checking blocks is overly specific

### ISSUE THREAD

**Issue by eli-schwartz** [snok/flake8-type-checking#68]
I just discovered this project and it looks really interesting and useful. However, I hit a slight snag.

Consider the following file:
```
$ cat foo.py 
import typing

if typing.TYPE_CHECKING:
    from pathlib import Path

p: Path
$ flake8 foo.py
foo.py:4:5: TC002 Move third-party import 'pathlib.Path' into a type-checking block
```

Same applies for if `T.TYPE_CHECKING`.

The README suggests using `from typing import TYPE_CHECKING`, and this does work, but I don't want that. It is a common pattern to use names such as `T.Any`, and guard annotation imports from your own project only, and I have a lot of code which happily does that.

Oh, and on that note, a similar example:
```
$ cat foo.py 
import typing as T

l: T.List[str]
$ flake8 foo.py
foo.py:1:1: TC002 Move third-party import 'T' into a type-checking block
```

This one import for the typing module itself should probably be exempt, as `typing` or `T` is the only thing you cannot restrict yourself to importing inside a T.TYPE_CHECKING block.

**Comment by sondrelg** [comment]
Thanks for opening the issue 👏 

You're right about the `TYPE_CHECKING` block detection - it could definitely be improved. Would you be up for trying to create a fix for this @eli-schwartz?

> It is a common pattern to ... guard annotation imports from your own project only

Just in case it's helpful and you're not aware of it, there is a `type-checking-exempt-modules` setting you can use to ignore typing altogether.

With respect to which types of imports to guard - this plugin makes an attempt at drawing a distinction between local and third-party imports, so you could also ignore `TC002` completely if you're using it mostly for helping with import circularity issues. I've been meaning to try to rewrite the import classifications using [this](https://github.com/asottile/aspy.refactor_imports#classify-a-module) library, but ran into some issue last time I looked at it 🙂

**Comment by sondrelg** [comment]
Hi again @eli-schwartz. I've added support for the following patterns:

```
from typing import TYPE_CHECKING

if TYPE_CHECKING
```

```
import typing 

if typing.TYPE_CHECKING
```

```
from typing import TYPE_CHECKING as TC

if TC
```

```
import typing as T

if T.TYPE_CHECKING
```

In https://github.com/snok/flake8-type-checking/pull/73. Would you mind testing that the branch works for you? 

Let me know if you need help 👍

**Comment by eli-schwartz** [comment]
Actually the thirdparty distinction isn't very important to me, as it happens. :D https://github.com/mesonbuild/meson/ has a strict stdlib-only policy. So basically all typechecking types tend to either be project-specific types or else an import from the typing module (and in some rare cases a stdlib type used at runtime too).

> so you could also ignore `TC002` completely if you're using it mostly for helping with import circularity issues

Oddly enough, all of our imports seem to get classified as TC002 anyway, which is mentioned in the testsuite code comments as "semi-arbitrary". And they are done using `from .foo import bar` due to Reasons™ which is admittedly a bit odd. (OTOH in ImportVisitor._import_is_local you can presumably consider this a fourth assumption which is an ironclad one.)

...

Anyway, I have tried your branch with promising results:

```
$ flake8 mesonbuild/ | grep TC0
mesonbuild/build.py:41:1: TC002 Move third-party import 'compilers.Compiler' into a type-checking block
mesonbuild/build.py:45:1: TC002 Move third-party import 'linkers.StaticLinker' into a type-checking block
mesonbuild/coredata.py:282:1: TC003 Found multiple type checking blocks
mesonbuild/environment.py:58:1: TC003 Found multiple type checking blocks
mesonbuild/minstall.py:29:1: TC002 Move third-party import 'backend.backends.InstallEmptyDir' into a type-checking block
mesonbuild/minstall.py:29:1: TC002 Move third-party import 'backend.backends.InstallSymlinkData' into a type-checking block
mesonbuild/minstall.py:29:1: TC002 Move third-party import 'backend.backends.InstallDataBase' into a type-checking block
mesonbuild/minstall.py:29:1: TC002 Move third-party import 'backend.backends.TargetInstallData' into a type-checking block
mesonbuild/minstall.py:29:1: TC002 Move third-party import 'backend.backends.ExecutableSerialisation' into a type-checking block
mesonbuild/mintro.py:29:1: TC002 Move third-party import 'mparser.BaseNode' into a type-checking block
mesonbuild/mintro.py:30:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/mintro.py:34:1: TC002 Move third-party import 'argparse' into a type-checking block
mesonbuild/msubprojects.py:1:1: TC002 Move third-party import 'dataclasses.InitVar' into a type-checking block
mesonbuild/msubprojects.py:16:1: TC002 Move third-party import 'wrap.wrap.PackageDefinition' into a type-checking block
mesonbuild/rewriter.py:30:1: TC002 Move third-party import 'mparser.BaseNode' into a type-checking block
mesonbuild/ast/interpreter.py:18:1: TC002 Move third-party import 'visitor.AstVisitor' into a type-checking block
mesonbuild/ast/interpreter.py:22:1: TC002 Move third-party import 'interpreterbase.TYPE_nvar' into a type-checking block
mesonbuild/ast/interpreter.py:22:1: TC002 Move third-party import 'interpreterbase.TYPE_nkwargs' into a type-checking block
mesonbuild/ast/interpreter.py:42:1: TC002 Move third-party import 'mparser.AndNode' into a type-checking block
mesonbuild/ast/interpreter.py:42:1: TC002 Move third-party import 'mparser.IndexNode' into a type-checking block
mesonbuild/ast/interpreter.py:42:1: TC002 Move third-party import 'mparser.OrNode' into a type-checking block
mesonbuild/ast/interpreter.py:42:1: TC002 Move third-party import 'mparser.IfClauseNode' into a type-checking block
mesonbuild/ast/interpreter.py:42:1: TC002 Move third-party import 'mparser.ForeachClauseNode' into a type-checking block
mesonbuild/ast/interpreter.py:42:1: TC002 Move third-party import 'mparser.UMinusNode' into a type-checking block
mesonbuild/ast/interpreter.py:42:1: TC002 Move third-party import 'mparser.ComparisonNode' into a type-checking block
mesonbuild/ast/introspection.py:19:1: TC002 Move third-party import 'visitor.AstVisitor' into a type-checking block
mesonbuild/ast/introspection.py:23:1: TC002 Move third-party import 'interpreterbase.TYPE_nvar' into a type-checking block
mesonbuild/ast/introspection.py:24:1: TC002 Move third-party import 'build.BuildTarget' into a type-checking block
mesonbuild/ast/postprocess.py:19:1: TC002 Move third-party import 'None.mparser' into a type-checking block
mesonbuild/ast/postprocess.py:20:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/ast/printer.py:21:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/ast/visitor.py:18:1: TC002 Move third-party import 'None.mparser' into a type-checking block
mesonbuild/backend/backends.py:16:1: TC002 Move third-party import 'dataclasses.InitVar' into a type-checking block
mesonbuild/backend/ninjabackend.py:50:1: TC002 Move third-party import 'build.ExtractedObjects' into a type-checking block
mesonbuild/backend/ninjabackend.py:51:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/backend/vs2010backend.py:28:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/backend/vs2012backend.py:17:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/backend/vs2012backend.py:18:1: TC002 Move third-party import 'build.Build' into a type-checking block
mesonbuild/backend/vs2012backend.py:19:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/backend/vs2013backend.py:17:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/backend/vs2013backend.py:18:1: TC002 Move third-party import 'build.Build' into a type-checking block
mesonbuild/backend/vs2013backend.py:19:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/backend/vs2015backend.py:17:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/backend/vs2015backend.py:18:1: TC002 Move third-party import 'build.Build' into a type-checking block
mesonbuild/backend/vs2015backend.py:19:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/backend/vs2017backend.py:21:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/backend/vs2017backend.py:22:1: TC002 Move third-party import 'build.Build' into a type-checking block
mesonbuild/backend/vs2019backend.py:20:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/backend/vs2019backend.py:21:1: TC002 Move third-party import 'build.Build' into a type-checking block
mesonbuild/backend/vs2022backend.py:20:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/backend/vs2022backend.py:21:1: TC002 Move third-party import 'build.Build' into a type-checking block
mesonbuild/backend/xcodebackend.py:24:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/cmake/executor.py:19:1: TC002 Move third-party import 'pathlib.Path' into a type-checking block
mesonbuild/cmake/executor.py:26:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/cmake/generator.py:16:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/cmake/interpreter.py:18:1: TC002 Move third-party import 'common.TargetOptions' into a type-checking block
mesonbuild/cmake/interpreter.py:18:1: TC002 Move third-party import 'common.CMakeConfiguration' into a type-checking block
mesonbuild/cmake/interpreter.py:22:1: TC002 Move third-party import 'traceparser.CMakeGeneratorTarget' into a type-checking block
mesonbuild/cmake/toolchain.py:18:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/c.py:20:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/compilers.py:25:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/cpp.py:22:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/cpp.py:24:1: TC002 Move third-party import 'compilers.CompileCheckMode' into a type-checking block
mesonbuild/compilers/cs.py:22:1: TC002 Move third-party import 'compilers.MachineChoice' into a type-checking block
mesonbuild/compilers/cuda.py:22:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/cuda.py:26:1: TC002 Move third-party import 'compilers.CompileCheckMode' into a type-checking block
mesonbuild/compilers/d.py:20:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/detect.py:15:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/detect.py:22:1: TC002 Move third-party import 'linkers.StaticLinker' into a type-checking block
mesonbuild/compilers/detect.py:53:1: TC002 Move third-party import 'compilers.Compiler' into a type-checking block
mesonbuild/compilers/detect.py:54:1: TC002 Move third-party import 'c.CCompiler' into a type-checking block
mesonbuild/compilers/detect.py:76:1: TC002 Move third-party import 'cpp.CPPCompiler' into a type-checking block
mesonbuild/compilers/detect.py:104:1: TC002 Move third-party import 'fortran.FortranCompiler' into a type-checking block
mesonbuild/compilers/detect.py:121:1: TC002 Move third-party import 'objc.ObjCCompiler' into a type-checking block
mesonbuild/compilers/detect.py:127:1: TC002 Move third-party import 'objcpp.ObjCPPCompiler' into a type-checking block
mesonbuild/compilers/detect.py:134:1: TC002 Move third-party import 'rust.RustCompiler' into a type-checking block
mesonbuild/compilers/fortran.py:33:1: TC002 Move third-party import 'mesonbuild.mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/java.py:22:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/objc.py:18:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/objcpp.py:18:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/rust.py:20:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/swift.py:18:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/vala.py:19:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/compilers/mixins/arm.py:28:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/ccrx.py:24:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/clike.py:47:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/compcert.py:23:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/emscripten.py:27:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/gnu.py:33:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/islinker.py:30:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/pgi.py:26:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/ti.py:24:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/visualstudio.py:30:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/compilers/mixins/xc16.py:24:1: TC004 Move import 'Compiler' out of type-checking block. Import is used for more than type hinting.
mesonbuild/dependencies/boost.py:22:1: TC002 Move third-party import 'environment.Environment' into a type-checking block
mesonbuild/dependencies/cmake.py:17:1: TC002 Move third-party import 'cmake.CMakeTarget' into a type-checking block
mesonbuild/dependencies/detect.py:37:1: TC003 Found multiple type checking blocks
mesonbuild/dependencies/dev.py:28:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/dependencies/factory.py:19:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/dependencies/factory.py:21:1: TC002 Move third-party import 'base.ExternalDependency' into a type-checking block
mesonbuild/dependencies/pkgconfig.py:16:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/interpreter/dependencyfallbacks.py:1:1: TC002 Move third-party import 'interpreterobjects.SubprojectHolder' into a type-checking block
mesonbuild/interpreter/dependencyfallbacks.py:9:1: TC002 Move third-party import 'interpreterbase.TYPE_nvar' into a type-checking block
mesonbuild/interpreter/dependencyfallbacks.py:9:1: TC002 Move third-party import 'interpreterbase.TYPE_nkwargs' into a type-checking block
mesonbuild/interpreter/interpreter.py:35:1: TC002 Move third-party import 'interpreterbase.baseobjects.TYPE_var' into a type-checking block
mesonbuild/interpreter/interpreter.py:35:1: TC002 Move third-party import 'interpreterbase.baseobjects.TYPE_kwargs' into a type-checking block
mesonbuild/interpreter/interpreter.py:38:1: TC002 Move third-party import 'backend.backends.Backend' into a type-checking block
mesonbuild/interpreter/interpreter.py:91:1: TC004 Move import 'kwargs' out of type-checking block. Import is used for more than type hinting.
mesonbuild/interpreter/interpreterobjects.py:16:1: TC002 Move third-party import 'interpreterbase.TYPE_var' into a type-checking block
mesonbuild/interpreter/interpreterobjects.py:16:1: TC002 Move third-party import 'interpreterbase.FeatureCheckBase' into a type-checking block
mesonbuild/interpreter/interpreterobjects.py:16:1: TC002 Move third-party import 'interpreterbase.TYPE_kwargs' into a type-checking block
mesonbuild/interpreter/interpreterobjects.py:16:1: TC002 Move third-party import 'interpreterbase.InterpreterObject' into a type-checking block
mesonbuild/interpreter/interpreterobjects.py:16:1: TC002 Move third-party import 'interpreterbase.TYPE_nvar' into a type-checking block
mesonbuild/interpreter/interpreterobjects.py:16:1: TC002 Move third-party import 'interpreterbase.TYPE_nkwargs' into a type-checking block
mesonbuild/interpreter/interpreterobjects.py:26:1: TC002 Move third-party import 'mesonlib.MesonException' into a type-checking block
mesonbuild/interpreter/interpreterobjects.py:31:1: TC004 Move import 'kwargs' out of type-checking block. Import is used for more than type hinting.
mesonbuild/interpreter/kwargs.py:9:1: TC002 Move third-party import 'typing_extensions.Literal' into a type-checking block
mesonbuild/interpreter/kwargs.py:11:1: TC002 Move third-party import 'None.build' into a type-checking block
mesonbuild/interpreter/kwargs.py:12:1: TC002 Move third-party import 'None.coredata' into a type-checking block
mesonbuild/interpreter/kwargs.py:13:1: TC002 Move third-party import 'mesonlib.OptionKey' into a type-checking block
mesonbuild/interpreter/kwargs.py:13:1: TC002 Move third-party import 'mesonlib.FileMode' into a type-checking block
mesonbuild/interpreter/kwargs.py:13:1: TC002 Move third-party import 'mesonlib.File' into a type-checking block
mesonbuild/interpreter/kwargs.py:13:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/interpreter/kwargs.py:13:1: TC002 Move third-party import 'mesonlib.FileOrString' into a type-checking block
mesonbuild/interpreter/kwargs.py:14:1: TC002 Move third-party import 'programs.ExternalProgram' into a type-checking block
mesonbuild/interpreter/type_checking.py:11:1: TC002 Move third-party import 'interpreterbase.TYPE_var' into a type-checking block
mesonbuild/interpreter/primitives/array.py:6:1: TC002 Move third-party import 'interpreterbase.TYPE_kwargs' into a type-checking block
mesonbuild/interpreter/primitives/boolean.py:4:1: TC002 Move third-party import 'interpreterbase.TYPE_var' into a type-checking block
mesonbuild/interpreter/primitives/boolean.py:4:1: TC002 Move third-party import 'interpreterbase.TYPE_kwargs' into a type-checking block
mesonbuild/interpreter/primitives/dict.py:6:1: TC002 Move third-party import 'interpreterbase.TYPE_kwargs' into a type-checking block
mesonbuild/interpreter/primitives/integer.py:4:1: TC002 Move third-party import 'interpreterbase.TYPE_var' into a type-checking block
mesonbuild/interpreter/primitives/integer.py:4:1: TC002 Move third-party import 'interpreterbase.TYPE_kwargs' into a type-checking block
mesonbuild/interpreter/primitives/string.py:10:1: TC002 Move third-party import 'interpreterbase.TYPE_var' into a type-checking block
mesonbuild/interpreter/primitives/string.py:10:1: TC002 Move third-party import 'interpreterbase.TYPE_kwargs' into a type-checking block
mesonbuild/interpreterbase/_unholder.py:15:1: TC002 Move third-party import 'baseobjects.TYPE_var' into a type-checking block
mesonbuild/interpreterbase/baseobjects.py:44:1: TC003 Found multiple type checking blocks
mesonbuild/interpreterbase/decorators.py:16:1: TC002 Move third-party import 'baseobjects.TYPE_kwargs' into a type-checking block
mesonbuild/interpreterbase/decorators.py:19:1: TC002 Move third-party import 'operator.MesonOperator' into a type-checking block
mesonbuild/interpreterbase/decorators.py:119:1: TC003 Found multiple type checking blocks
mesonbuild/interpreterbase/disabler.py:15:1: TC002 Move third-party import 'baseobjects.TYPE_var' into a type-checking block
mesonbuild/interpreterbase/disabler.py:15:1: TC002 Move third-party import 'baseobjects.TYPE_kwargs' into a type-checking block
mesonbuild/interpreterbase/disabler.py:16:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/interpreterbase/interpreterbase.py:21:1: TC002 Move third-party import 'baseobjects.SubProject' into a type-checking block
mesonbuild/interpreterbase/interpreterbase.py:21:1: TC002 Move third-party import 'baseobjects.TYPE_kwargs' into a type-checking block
mesonbuild/linkers/detect.py:15:1: TC002 Move third-party import 'mesonlib.MachineChoice' into a type-checking block
mesonbuild/linkers/detect.py:19:1: TC002 Move third-party import 'linkers.GnuDynamicLinker' into a type-checking block
mesonbuild/linkers/detect.py:19:1: TC002 Move third-party import 'linkers.DynamicLinker' into a type-checking block
mesonbuild/mesonlib/__init__.py:22:1: TC002 Move third-party import 'vsenv.setup_vsenv' into a type-checking block
mesonbuild/mesonlib/platform.py:20:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/mesonlib/posix.py:20:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/mesonlib/win32.py:20:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/modules/cmake.py:22:1: TC002 Move third-party import 'cmake.SingleTargetOptions' into a type-checking block
mesonbuild/modules/fs.py:22:1: TC002 Move third-party import 'mesonlib.FileOrString' into a type-checking block
mesonbuild/modules/gnome.py:33:1: TC002 Move third-party import 'build.BuildTarget' into a type-checking block
mesonbuild/modules/java.py:23:1: TC002 Move third-party import 'None.ModuleState' into a type-checking block
mesonbuild/modules/java.py:24:1: TC002 Move third-party import 'interpreter.Interpreter' into a type-checking block
mesonbuild/modules/python.py:626:1: TC003 Found multiple type checking blocks
mesonbuild/modules/qt.py:26:1: TC002 Move third-party import 'mesonlib.FileOrString' into a type-checking block
mesonbuild/modules/qt.py:30:1: TC002 Move third-party import 'programs.ExternalProgram' into a type-checking block
mesonbuild/modules/qt.py:36:1: TC004 Move import 'kwargs' out of type-checking block. Import is used for more than type hinting.
mesonbuild/modules/unstable_cuda.py:19:1: TC002 Move third-party import 'compilers.Compiler' into a type-checking block
mesonbuild/scripts/clangformat.py:21:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/clangtidy.py:20:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/cleantrees.py:19:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/coverage.py:18:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/delwithsuffix.py:16:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/depscan.py:23:1: TC002 Move third-party import 'backend.ninjabackend.TargetDependencyScannerInfo' into a type-checking block
mesonbuild/scripts/dirchanger.py:19:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/gettext.py:18:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/hotdochelper.py:8:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/meson_exe.py:20:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/msgfmthelper.py:18:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/regen_checker.py:17:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/run_tool.py:15:1: TC002 Move third-party import 'subprocess' into a type-checking block
mesonbuild/scripts/scanbuild.py:22:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/tags.py:18:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/uninstall.py:16:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/scripts/vcstagger.py:16:1: TC002 Move third-party import 'T' into a type-checking block
mesonbuild/templates/mesontemplates.py:15:1: TC002 Move third-party import 'argparse' into a type-checking block
mesonbuild/templates/samplefactory.py:25:1: TC002 Move third-party import 'mesonbuild.templates.sampleimpl.SampleImpl' into a type-checking block
mesonbuild/templates/samplefactory.py:27:1: TC002 Move third-party import 'argparse' into a type-checking block
```

After some spot-checking, these tend to invariably be correctly reported (and in several cases reveals a `kwargs` typing-related module which is also overridden repeatedly as a runtime variable, heh. We should have been clued in by functions which took the parameter `kwargs: kwargs.FooTypedDict`). The exception is some files which do `import typing as T`, then do not define an `if T.TYPE_CHECKING` block, but do use `T.List` or what have you in the annotations. From a quick look at your proposed implementation:

> ```
> # We could have just added typing as a blanket ignored module,
> # but that's a breaking change, so this will do for now.
> ```

So this seems to be expected, and as you mentioned above there is a fairly easy and reliable workaround. I *personally* would feel confident making it a blanket ignored module, but that may just be me. Adding it to `type-checking-exempt-modules = typing` is fine with me too.

tl;dr the branch seems to work great.

**Comment by sondrelg** [comment]
Thanks for testing, and thanks for the feedback. 

I'm a little tempted to redo the import classification using something like [this](https://github.com/asottile/aspy.refactor_imports#classify-a-module), and making `typing` exempt, and release a v2. I'll just need to find a day to do it I think. I like also putting typing imports into type checking blocks, because it feels more organized, but I'm pretty convinced that behavior shouldn't be the default 👍 

I'll release the fix shortly.
